### ISSUE TITLE
Rewrite QL using Gstreamer Editing Services

### ISSUE THREAD

**Issue by lazka** [quodlibet/quodlibet#801]
Original [issue 801](https://code.google.com/p/quodlibet/issues/detail?id=801) created by LRN1986 on 2011-08-21T14:16:12.000Z:

GES is the new API that makes it easy to write audio/video editors using GStreamer. GStreamer already provided GNonLin before, but GNonLin is somewhat tricky to use. GES, on the other hand, focuses on the ease of use.

Why is it better than playbin/decodebin:
1) All kinds of audio filters can be easily applied in a generic fashion.
2) It's easy to do partial playback (yes, the famous &quot;wontfix&quot; CUE sheet support, as well as playing bookmarked/chaptered video files as separate music tracks - all of that will be implementable)
3) Gaplessness can be easily implemented in QL in the way that QL/user wants, instead of relying on playbin to do the right thing.
4) It should be possible to implement crossfade (which is, AFAIK, a &quot;wontfix&quot; at the moment).

How to do:
Basically, make a GES Timeline pipeline, connect it with the audio sink via queue. Populate the timeline with audio clips (with appropriate effects, if needed) and start playing. At some point during the playtime evaluate the timeline condition: if it is running out of audio clips, pause the timeline pipeline (the queue will keep feeding the audio sink with data for some time), add enough audio clips (again, with appropriate effects) to be safe, remove the clips that are already played out, resume the timeline pipeline.
That's the theory.

**Comment by lazka** [comment]
atm nothing is using ges besides pitivi. If someone implements an alternative audio backend I'd be happy to include it (and also make backend switching easier if needed)
